# Проект по SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Цели исследования**

Проанализировать базу данных и предоставить данные, которые помогут сформулировать ценностное предложение для нового приложения для тех, кто любит читать.

**Описание данных**

***Таблица books***

Содержит данные о книгах:
* book_id — идентификатор книги;
* author_id — идентификатор автора;
* title — название книги;
* num_pages — количество страниц;
* publication_date — дата публикации книги;
* publisher_id — идентификатор издателя.

***Таблица authors***

Содержит данные об авторах:
* author_id — идентификатор автора;
* author — имя автора.

***Таблица publishers***

Содержит данные об издательствах:
* publisher_id — идентификатор издательства;
* publisher — название издательства;

***Таблица ratings***

Содержит данные о пользовательских оценках книг:
* rating_id — идентификатор оценки;
* book_id — идентификатор книги;
* username — имя пользователя, оставившего оценку;
* rating — оценка книги.

***Таблица reviews***

Содержит данные о пользовательских обзорах:
* review_id — идентификатор обзора;
* book_id — идентификатор книги;
* username — имя автора обзора;
* text — текст обзора.

**Схема данных**

https://pictures.s3.yandex.net/resources/scheme_1589269096.png

**Задания**
* Посчитайте, сколько книг вышло после 1 января 2000 года;
* Для каждой книги посчитайте количество обзоров и среднюю оценку;
* Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
* Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
* Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

**Как выполнить задание?**
* Опишите цели исследования;
* Исследуйте таблицы — выведите первые строки;
* Сделайте по одному SQL-запросу для решения каждого задания;
* Выведите результаты каждого запроса в тетрадке;
* Опишите выводы по каждой из решённых задач.

**Примечания**
* Не забывайте про функции, они упрощают работу;
* Получите ответы, используя только язык запросов SQL. Pandas допускается только для хранения и вывода результатов.

## Загрузка данных и изучение общей информаии

In [1]:
# Импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
# Устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# Сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
# Чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [2]:
# Выведем таблицу authors
query = '''SELECT * FROM authors LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [3]:
# Выведем таблицу publishers
query = '''SELECT * FROM publishers LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [4]:
# Выведем таблицу ratings
query = '''SELECT * FROM ratings LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [5]:
# Выведем таблицу reviews
query = '''SELECT * FROM reviews LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

## Выполнение заданий

### Задание №1. Посчитайте, сколько книг вышло после 1 января 2000 года.

In [6]:
query = '''
SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01';
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0    819

**Вывод:**

После 1 января 2000 года вышло 819 книг.

### Задание №2. Для каждой книги посчитайте количество обзоров и среднюю оценку

In [7]:
query = '''
SELECT b.title, 
       COUNT(DISTINCT rev.review_id) AS review_count, 
       AVG(ra.rating) AS avg_rating
FROM books AS b
LEFT JOIN reviews AS rev ON b.book_id = rev.book_id
LEFT JOIN ratings AS ra ON b.book_id = ra.book_id
GROUP BY b.book_id
ORDER BY avg_rating DESC,
         review_count DESC;
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

title  review_count  \
0                        A Dirty Job (Grim Reaper  #1)             4   
1              School's Out—Forever (Maximum Ride  #2)             3   
2         Moneyball: The Art of Winning an Unfair Game             3   
3        Arrows of the Queen (Heralds of Valdemar  #1)             2   
4    Wherever You Go  There You Are: Mindfulness Me...             2   
..                                                 ...           ...   
995  The World Is Flat: A Brief History of the Twen...             3   
996                                      Drowning Ruth             3   
997                  His Excellency: George Washington             2   
998                                              Junky             2   
999                               Harvesting the Heart             2   

     avg_rating  
0          5.00  
1          5.00  
2          5.00  
3          5.00  
4          5.00  
..          ...  
995        2.25  
996        2.00  
997        2.00  
998        2.00  
999        1.50  

[1000 rows x 3 columns]

**Вывод:**

Для каждой книги было посчитано количество обзоров и средняя оценка. Максимальная оценка - 5.0, минимальная - 1.5. На первом месте книга A Dirty Job (Grim Reaper #1) с 4 обзорами на нее и средней оценкой 5.0, на последнем - книга Harvesting the Heart с 2 обзорами на нее и средней оценкой 1.5.

### Задание №3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [8]:
query = '''
SELECT p.publisher, 
       COUNT(DISTINCT b.book_id) AS books_count
FROM publishers AS p
LEFT JOIN books AS b ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY books_count DESC
LIMIT 3;
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)


publisher  books_count
0             Penguin Books           42
1                   Vintage           31
2  Grand Central Publishing           25

**Вывод:**

Для каждого издательства было посчитано количество книг толще 50 страниц. Издательство, которое выпустило наибольшее число книг толще 50 страниц - Penguin Books с 42 книгами. На втором месте - издательство Vintage с 31 книгой, на третьем месте - Grand Central Publishing с 25 книгами.

### Задание №4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [9]:
query = '''
SELECT a.author, 
       AVG(ra.rating) AS avg_rating
FROM authors AS a
LEFT JOIN books AS b ON b.author_id = a.author_id
LEFT JOIN ratings AS ra ON ra.book_id = b.book_id
WHERE ra.book_id IN (SELECT ra.book_id
                     FROM ratings AS ra
                     GROUP BY ra.book_id
                     HAVING COUNT(ra.rating_id) >= 50)
GROUP BY a.author
ORDER BY avg_rating DESC
LIMIT 3;
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)


author  avg_rating
0         J.K. Rowling/Mary GrandPré    4.287097
1  Markus Zusak/Cao Xuân Việt Khương    4.264151
2                     J.R.R. Tolkien    4.246914

**Вывод:**

Для каждого автора была посчитана средняя оценка книг. Автор, который получил самую высокую среднюю оценку книг, - J.K. Rowling/Mary GrandPré с оценкой 4.28. На втором месте - Markus Zusak/Cao Xuân Việt Khương с оценкой 4.26, на третьем месте - J.R.R. Tolkien с оценкой 4.24.

### Задание №5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [10]:
query = '''
WITH a AS (SELECT rev.username, 
                  COUNT(rev.review_id) AS review_count
           FROM reviews AS rev
           WHERE rev.username IN (SELECT r.username
                                  FROM ratings AS r
                                  GROUP BY r.username
                                  HAVING COUNT(r.rating_id) > 48)
           GROUP BY rev.username)
SELECT AVG(review_count) AS avg_review_count 
FROM a
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

avg_review_count
0              24.0

**Вывод:**

Среднее количество обзоров от пользователей, которые поставили больше 48 оценок, - 24.

## Вывод

* После 1 января 2000 года вышло 819 книг.
* Для каждой книги было посчитано количество обзоров и средняя оценка. Максимальная оценка - 5.0, минимальная - 1.5. На первом месте книга A Dirty Job (Grim Reaper #1) с 4 обзорами на нее и средней оценкой 5.0, на последнем - книга Harvesting the Heart с 2 обзорами на нее и средней оценкой 1.5.
* Для каждого издательства было посчитано количество книг толще 50 страниц. Издательство, которое выпустило наибольшее число книг толще 50 страниц - Penguin Books с 42 книгами. На втором месте - издательство Vintage с 31 книгой, на третьем месте - Grand Central Publishing с 25 книгами.
* Для каждого автора была посчитана средняя оценка книг. Автор, который получил самую высокую среднюю оценку книг, - J.K. Rowling/Mary GrandPré с оценкой 4.28. На втором месте - Markus Zusak/Cao Xuân Việt Khương с оценкой 4.26, на третьем месте - J.R.R. Tolkien с оценкой 4.24.
* Среднее количество обзоров от пользователей, которые поставили больше 48 оценок, - 24.